src/loop/parser.ml

```diff
@@ -190,70 +200,98 @@ module Make(State : State.S) = struct
-            st, file, gen_of_llist l
-          end
-        | `File f ->
-          let s = Dolmen.Std.Statement.include_ f [] in
-          (* Auto-detect input format *)
-          let lang =
-            match file.lang with
-            | Some l -> l
-            | None ->
-              let res, _, _ = Logic.of_filename f in
-              res
-          in
-          (* Formats Dimacs and Tptp are descriptive and lack the emission
-              of formal solve/prove instructions, so we need to add them. *)
-          let s' =
-            match lang with
-            | Logic.Zf
-            | Logic.ICNF
-            | Logic.Smtlib2 _
-            | Logic.Alt_ergo -> s
-            | Logic.Dimacs
-            | Logic.Tptp _ ->
-              Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
+          st, file, gen_finally gen cl
+        | Some `Full ->
+          let lang, file_loc, l = Logic.parse_raw_lazy ~language:lang
+            ~filename contents
           in
-          let file = { file with lang = Some lang; } in
-          st, file, (Gen.singleton s')
-      with
-      | Logic.Extension_not_found ext ->
-        State.error st extension_not_found ext, file, Gen.empty
-    in
-    let st = set_logic_file st file in
-    (* Wrap the resulting parser *)
-    st, wrap_parser ~file (Gen.append (Gen.of_list prelude) g)
+          let file = { file with loc = file_loc; lang = Some lang; } in
+          st, file, gen_of_llist l
+        end
+      | `File f ->
+        let s = Dolmen.Std.Statement.include_ f [] in
+        let lang = Option.get file.lang in
+        (* Formats Dimacs and Tptp are descriptive and lack the emission
+            of formal solve/prove instructions, so we need to add them. *)
+        let s' =
+          match lang with
+          | Logic.Zf
+          | Logic.ICNF
+          | Logic.Smtlib2 _
+          | Logic.Alt_ergo -> s
+          | Logic.Dimacs
+          | Logic.Tptp _ ->
+            Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
+        in
+        let file = { file with lang = Some lang; } in
+        st, file, (Gen.singleton s')
+    | exception Logic.Extension_not_found ext ->
+      State.error st extension_not_found ext, file, Gen.empty
+
+  exception Stdin_prelude of Logic.language file
+
+  let parse_logic ?(preludes = []) prelude st file =
+    try
+      let st, gp =
+        List.fold_left (fun (st, gp) (file : _ file) ->
+          match file.source with
+          | `Stdin -> raise (Stdin_prelude file)
+          | _ ->
+            match infer_lang file with
+            | file ->
+              let st, file = switch_to_full_mode_if_needed st file file in
+              let st, file, g = parse_file st file in
+              st, wrap_append ~file gp (wrap_parser ~file g)
+            | exception Logic.Extension_not_found ext ->
+              State.error st extension_not_found ext, gp
+        ) (st, (fun st -> st, None)) preludes
```

Comment:
Unless I missed something, this fold will make it so that the files are parsed beginning from the last element of the list of prelude, and parsing the first element of the list last, which is the inverse of what i'd expect.
This transformation might be easier by doing something like the following pseudo-code:
```
preludes |> to_gen |> Gen.flat_map (fun file -> wrap_gen (set_logic_file ..) (parse_file file))
```

Reply:
Right, the `parse_file` calls are performed in the reverse order, but the resulting generator should be in the right order.

Using `Gen` directly is probably better, yes, I am just not used to it — will change it to your suggestion

Reply:
Actually re-reading the code I think it is correct. `parse_file` doesn't actually parse, it returns a lazy generator (unless I missed something, even in the `Full` case, it is a `S.t list Lazy.t`). `Gen.flat_map` doesn't work because the outer state needs to be threaded through. With `List.fold_left` the outer state is propagated from beginning to end and the included list is built in the right order (note `wrap_append ~file gp (wrap_parser ~file g)` and not `wrap_append ~file (wrap_parser ~file g) gp` which would actually be incorrect due to the definition of `wrap_append`).

(As a side note, `wrap_append` is really a poor name, because it `appends` and then changes the file; open to suggestions)

Reply:
Alright, so after re-reading it, the code is indeed correct, but it's a bit of a headache to make sure.

I don't see the problem with `Gen.flat_map` : it's not really necessary to thread, or even have an "outer" state: it's perfectly fine to change the state only when we begin to parse it (and change it on the fly during the pipeline, instead of at the beginning).

Reply:
Yeah, if we don't need to thread the outer state, then `flat_map` is easier for sure (and we probably can get rid of `wrap_append` entirely). Will change

Reply:
Great ! (just for the reference: the "outer" state is mainly here for historical reasons: at the beginning the generators returned for the parsed statements used to be a simple `unit -> parsed_statement` function, and didn't thread the state around. I later fixed that to pass the state around, but left a lot of code that actually updated the outer state)